lib/srv/reexec.go

```diff
@@ -761,42 +660,241 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
+	// parentConn is a datagram Unix socket, which is not connection oriented
+	// and thus won't unblock when the parent closes its side of the connection.
+	// Instead we use an interrupt signal from the parent process to unblock.
+	intC := make(chan os.Signal, 1)
+	signal.Notify(intC, os.Interrupt)
+	go func() {
+		<-intC
+		parentConn.Close()
+	}()
+
 	for {
 		buf := make([]byte, 1024)
 		fbuf := make([]*os.File, 1)
-		n, fn, err := uds.ReadWithFDs(conn, buf, fbuf)
+		n, fn, err := uds.ReadWithFDs(parentConn, buf, fbuf)
 		if err != nil {
 			if utils.IsOKNetworkError(err) {
-				return errorWriter, teleport.RemoteCommandSuccess, nil
+				// parent connection closed, process should exit.
+				return errorWriter, teleport.RemoteCommandFailure, nil
 			}
-			return errorWriter, teleport.RemoteCommandFailure, trace.Wrap(err)
+			slog.ErrorContext(ctx, "Error reading networking request from parent.", "error", err)
+			continue
 		}
-		addr := string(buf[:n])
+
 		if fn == 0 {
-			log.Errorf("Parent did not send a file descriptor for address %q.", addr)
+			slog.ErrorContext(ctx, "Networking request requires a control file.")
+			continue
+		}
+
+		controlConn, err := uds.FromFile(fbuf[0])
+		_ = fbuf[0].Close()
+		if err != nil {
+			slog.ErrorContext(ctx, "Failed to get a connection from control file.")
 			continue
 		}
 
-		go func() {
-			if err := handler(ctx, addr, fbuf[0]); err != nil {
-				log.WithError(err).Errorf("Error handling forwarding request for address %q.", addr)
+		var req networking.Request
+		if err := json.Unmarshal(buf[:n], &req); err != nil {
+			slog.ErrorContext(ctx, "Error parsing networking request.", "error", err)
+			continue
+		}
+
+		// Some PAM modules (e.g. pam_namespace) do not behave properly in multithreaded contexts.
+		// Therefore we favor handling requests and cleanup on the main PAM thread for requests that
+		// are expected to be impacted (unix socket listeners).
+		switch req.Operation {
+		case networking.NetworkingOperationDial, networking.NetworkingOperationListen:
+			switch req.Network {
+			case "tcp":
+				// There are currently no known issues with tcp listen/dial in a multithreaded PAM context.
+				go func() {
+					cleanup := handleNetworkingRequest(ctx, controlConn, req)
+					if cleanup != nil {
+						context.AfterFunc(ctx, func() {
```

Comment:
This (and the defer later) are 100% not going to work, they will accumulate memory for each connection for the lifetime of the process even for short lived connections.

If we need to keep a list of paths to be deleted it should be an explicit list of paths and we should probably have some limit around how many we end up dealing with.

Reply:
This cleanup is only for agent/x11/remote forwarding _unix_ listeners, not local forwarding connections or remote forwarding tcp listeners. IMO the only thing missing here might be handling "cancel-tcpip-forward" for _unix listeners which we don't support_. As it currently stands each networking process will defer at most 2 cleanups, one for x11 and one for agent forwarding. 

I do agree that handling this more explicitly is a good idea, with some stored list of sockets to close, I just am not convinced of its necessity at this time.

Reply:
The main issue is that the cleanup will run at the end of the networking process, not when the listener is closed. And wasn't there an issue we recently fixed because we assumed that there was only going to be one X11 forward request in the lifetime of a connection and there most definitely wasn't?

If we are always supposed to unlink the listener when it's closed, maybe we can just keep the per-request socket open on both sides, and close it after closing the listener, to signal the child to close it?

A really cursed idea for the future: when we pass back the unix listener, we also pass a file descriptor for the directory that contains it, and then no matter the namespace we can `unlinkat` the socket after closing it, from the parent process 👀